### PR TITLE
[ASTMatcher] Fix redundant macro expansion checks in getExpansionLocOfMacro

### PR DESCRIPTION
**Pull request by mandymimi**
A performance issue was descibed in #114521 

**Root Cause**: The function getExpansionLocOfMacro is responsible for finding the expansion location of macros. When dealing with macro parameters, it recursively calls itself to check the expansion of macro arguments. This recursive logic redundantly checks previous macro expansions, leading to significant performance degradation when macros are heavily nested.
Solution
**Modification**: Track already processed macros during recursion.
Implementation Details:
Introduced a data structure to record processed macros.
Before each recursive call, check if the macro has already been processed to avoid redundant calculations.
**Testing**: 
1. refer to #114521 
Finder->addMatcher(expr(isExpandedFromMacro("NULL")).bind("E"), this);
run clang-tidy on freecad/src/Mod/Path/App/AreaPyImp.cpp 
clang-tidy src/Mod/Path/App/AreaPyImp.cpp -checks=-*,testchecker -p=build/compile_commands.json     
checker runs normally
2. check-clang-unit pass

### CODE REVIEW

File: clang/lib/ASTMatchers/ASTMatchersInternal.cpp [hunk 1]
```diff
@@ -697,27 +698,61 @@ static bool isTokenAtLoc(const SourceManager &SM, const LangOptions &LangOpts,
   return !Invalid && Text == TokenText;
 }
 
-std::optional<SourceLocation>
-getExpansionLocOfMacro(StringRef MacroName, SourceLocation Loc,
-                       const ASTContext &Context) {
+namespace {
+struct SourceLocationHash {
+  std::size_t operator()(const SourceLocation &Loc) const {
```
Comment (AaronBallman) [inline clang/lib/ASTMatchers/ASTMatchersInternal.cpp:703]:
```suggestion
  std::size_t operator()(SourceLocation Loc) const {
```
A `SourceLocation` is a 32-bit value, so no need to pass by const reference. The same change can be made elsewhere in the patch.

File: clang/lib/ASTMatchers/ASTMatchersInternal.cpp [hunk 2]
```diff
@@ -36,6 +36,7 @@
 #include <cstddef>
 #include <optional>
 #include <string>
+#include <unordered_set>
```
Comment (AaronBallman) [inline clang/lib/ASTMatchers/ASTMatchersInternal.cpp:39]: Would `llvm::DenseSet` make more sense?

https://llvm.org/docs/ProgrammersManual.html#llvm-adt-denseset-h

File: clang/lib/ASTMatchers/ASTMatchersInternal.cpp [hunk 3]
```diff
@@ -697,27 +698,61 @@ static bool isTokenAtLoc(const SourceManager &SM, const LangOptions &LangOpts,
...
+static std::optional<SourceLocation> getExpansionLocOfMacroRecursive(
+    StringRef MacroName, SourceLocation Loc, const ASTContext &Context,
+    std::unordered_set<SourceLocation, SourceLocationHash, SourceLocationEqual>
+        &CheckedLocations) {
   auto &SM = Context.getSourceManager();
   const LangOptions &LangOpts = Context.getLangOpts();
   while (Loc.isMacroID()) {
+    if (CheckedLocations.count(Loc)) {
+      return std::nullopt;
+    }
```
Comment (AaronBallman) [inline clang/lib/ASTMatchers/ASTMatchersInternal.cpp:725]:
```suggestion
    if (CheckedLocations.count(Loc))
      return std::nullopt;
```
We have very odd coding style rules, sorry. :-)

https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements

File: clang/lib/ASTMatchers/ASTMatchersInternal.cpp [hunk 4]
```diff
@@ -697,27 +698,61 @@ static bool isTokenAtLoc(const SourceManager &SM, const LangOptions &LangOpts,
...
+                                              Expansion.getSpellingLoc(),
+                                              Context, CheckedLocations)) {
         return ArgLoc;
+      }
+    }
     Loc = Expansion.getExpansionLocStart();
-    if (isTokenAtLoc(SM, LangOpts, MacroName, Loc))
+    if (isTokenAtLoc(SM, LangOpts, MacroName, Loc)) {
       return Loc;
+    }
```
Comment (AaronBallman) [inline clang/lib/ASTMatchers/ASTMatchersInternal.cpp:742]:
```suggestion
    if (isTokenAtLoc(SM, LangOpts, MacroName, Loc))
      return Loc;
```

**Review by AaronBallman — COMMENTED** [review]
Thank you for working on a fix for this! Please be sure to update `clang/docs/ReleaseNotes.rst` so users know about the fix.

Test coverage for the changes is interesting in that I'm not certain there's a good test to add for this (the test would be one which is pretty expensive to run, I'd imagine). But if you can think of some good coverage to add to `clang/unittests/ASTMatchers/`, that'd be good to add as well.

File: clang/lib/ASTMatchers/ASTMatchersInternal.cpp (hunk 2 above)
Comment (mandymimi) [reply clang/lib/ASTMatchers/ASTMatchersInternal.cpp]: very good sugguestion! updated in f3f08877f08c

File: clang/lib/ASTMatchers/ASTMatchersInternal.cpp (hunk 3 above)
Comment (mandymimi) [reply clang/lib/ASTMatchers/ASTMatchersInternal.cpp]: updated in f3f08877f08c

File: clang/lib/ASTMatchers/ASTMatchersInternal.cpp (hunk 4 above)
Comment (mandymimi) [reply clang/lib/ASTMatchers/ASTMatchersInternal.cpp]: updated in f3f08877f08c

File: clang/lib/ASTMatchers/ASTMatchersInternal.cpp (hunk 1 above)
Comment (mandymimi) [reply clang/lib/ASTMatchers/ASTMatchersInternal.cpp]: Got it, thank you

**Review by AaronBallman — APPROVED** [review]
LGTM!

**Pull request merged**